File: src/main/java/j/SingleLinkedList.java
package j;

public class SingleLinkedList<T> {
    private final Node<T> head = new Node<>(null);
    private int size;

    public int getSize() {
        return this.size;
    }

    public void append(T value) {
        final Node<T> node;
        if (this.size == 0) {
            node = this.head;
        } else {
            node = this.getNodeAt(this.size - 1);
        }

        node.next = new Node<>(value);

        this.size++;
    }

    public void appendLeft(T value) {
        final Node<T> node = new Node<>(value);

        node.next = this.head.next;
        this.head.next = node;

        this.size++;
    }

    public void insert(int index, T value) {
        if (index == 0) {
            this.appendLeft(value);
        } else if (index == this.size) {
            this.append(value);
        } else {
            final Node<T> prevNode = this.getNodeAt(index - 1);
            final Node<T> node = new Node<>(value);

            node.next = prevNode.next;
            prevNode.next = node;

            this.size++;
        }
    }

    public T get(int index) {
        return this.getNodeAt(index).value;
    }

    public T removeAt(int index) {
        this.checkIndex(index);

        final Node<T> prevNode;
        if (index == 0) {
            prevNode = this.head;
        } else {
            prevNode = this.getNodeAt(index - 1);
        }

        final Node<T> node = prevNode.next;
        prevNode.next = node.next;

        this.size--;

        return node.value;
    }

    private Node<T> getNodeAt(int index) {
        this.checkIndex(index);

        Node<T> node = this.head;
        for (int i = 0; i < index + 1; i++) {
            node = node.next;
        }

        return node;
    }

    private void checkIndex(int index) {
        if (index < 0 || this.size <= index) {
            throw new IndexOutOfBoundsException(index);
        }
    }

    private static class Node<T> {
        T value;
        Node<T> next;

        public Node(T value) {
            this.value = value;
        }
    }
}

File: src/main/kotlin/k/SingleLinkedList.kt
package k

class SingleLinkedList<T> {
    private val head: Node<*> = Node(null)

    var size: Int = 0
        private set

    fun append(value: T) {
        val node: Node<*> = if (this.size == 0) {
            this.head
        } else {
            this.getNodeAt(this.size - 1);
        }

        node.next = Node(value = value)

        this.size++
    }

    fun appendLeft(value: T) {
        val node: Node<T> = Node(value = value)

        node.next = this.head.next
        this.head.next = node

        this.size++
    }

    fun insert(
        index: Int,
        value: T
    ) {
        when (index) {
            0 -> {
                this.appendLeft(value = value)
            }

            this.size -> {
                this.append(value = value)
            }

            else -> {
                val prevNode: Node<*> = this.getNodeAt(index = index - 1)
                val node: Node<T> = Node(value = value)

                node.next = prevNode.next
                prevNode.next = node

                this.size++
            }
        }
    }

    @Throws(IndexOutOfBoundsException::class)
    fun get(index: Int): T {
        return this.getNodeAt(index = index).value
    }

    @Throws(IndexOutOfBoundsException::class)
    fun removeAt(index: Int): T {
        this.checkIndex(index = index)

        val prevNode: Node<*> = if (index == 0) {
            this.head
        } else {
            this.getNodeAt(index = index - 1)
        }

        @Suppress("UNCHECKED_CAST")
        val node: Node<T> = prevNode.next as Node<T>
        prevNode.next = node.next

        this.size--

        return node.value
    }

    private fun getNodeAt(index: Int): Node<T> {
        this.checkIndex(index = index)

        var node: Node<*> = this.head
        (0 until index + 1).forEach { _: Int ->
            node = node.next!!
        }

        @Suppress("UNCHECKED_CAST")
        return node as Node<T>
    }

    @Throws(IndexOutOfBoundsException::class)
    private fun checkIndex(index: Int) {
        assert(value = index in 0 until this.size) {
            throw IndexOutOfBoundsException(index)
        }
    }

    private data class Node<T>(val value: T) {
        var next: Node<*>? = null
    }
}
